### PR TITLE
Expose Special issue types endpoint throughout the BIP API service

### PR DESCRIPTION
**Pull request by nelsestu**
The EE team has requested that the special_issue_types endpoint be exposed via a new rabbitmq queue on the bip exchange. This PR introduces many of the necessary components towards making that endpoint available for EE consumption. Accompanying the real bip data, is a mock implementation that can be used for local development.

## What was the problem?
special_issue_types were needed.

Associated tickets or Slack threads:
- #3213

## How does this fix it?[^1]
Adds the special_issue_types endpoint to all the necessary components of the bip service.

## How to test this PR
- submit a rabbitmq request to the newly create getSpecialIssueTypesQueue

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).

### CODE REVIEW

File: mocks/mock-bip-claims-api/src/main/java/gov/va/vro/mockbipclaims/model/bip/response/SpecialIssueTypesResponse.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import gov.va.vro.mockbipclaims.model.bip.ProviderResponse;
+import gov.va.vro.mockbipclaims.model.bip.SpecialIssueType;
+import lombok.Data;
+
+import java.util.List;
+
+/** SpecialIssueTypesResponse. */
+@Data
+public class SpecialIssueTypesResponse extends ProviderResponse {
+  private SpecialIssueType[] codeNamePairs;
```
Comment (dfitchett) [inline mocks/mock-bip-claims-api/src/main/java/gov/va/vro/mockbipclaims/model/bip/response/SpecialIssueTypesResponse.java:10]: I'd probably go with `List<SpecialIssueType>` to allow for using streams

File: q
Comment (dfitchett) [inline q:1]: Oooh whats this?

File: svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/GetSpecialIssueTypesResponse.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  // concurrency concern given present use case
+  private final Supplier<SpecialIssueType[]> allTypesSupplier;
+
+  public GetSpecialIssueTypesResponse(Supplier<SpecialIssueType[]> allTypesSupplier) {
+    super(BipPayloadResponse.builder());
+    this.allTypesSupplier = allTypesSupplier;
+  }
+
+  public SpecialIssueType[] get() {
+    if (ArrayUtils.isEmpty(codeNamePairs)) codeNamePairs = allTypesSupplier.get();
```
Comment (brostk) [inline svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/GetSpecialIssueTypesResponse.java]: For my own awareness, does the team have a convention on single-line if blocks?

**Review by brostk — APPROVED** [review]
This all looks good to me.  The mock enablement is great and the tests look like they should pass just fine.  One nitpick on GetSpecialIssuesTypeResponse but not a blocker.

File: svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/GetSpecialIssueTypesResponse.java (hunk 2 above)
Comment (dfitchett) [reply svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/GetSpecialIssueTypesResponse.java]: I guess not. I would have thought that spotlessJava wouldn't like it. I'd prefer not for readability.

File: svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/ParameterlessBipRequest.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Getter
+@Builder
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@NoArgsConstructor
+public class ParameterlessBipRequest implements BipPayloadRequest {}
```
Comment (dfitchett) [inline svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/ParameterlessBipRequest.java]: This class should prob be final!

File: mocks/mock-bip-claims-api/src/main/java/gov/va/vro/mockbipclaims/controller/ContentionsController.java [hunk 4]
```diff
@@ -141,8 +145,10 @@ public ResponseEntity<UpdateContentionsResponse> updateContentions(
 
   /** Not fully implemented. Only used for connectivity testing. */
```
Comment (lisac) [inline mocks/mock-bip-claims-api/src/main/java/gov/va/vro/mockbipclaims/controller/ContentionsController.java]:
```suggestion
  /** Responds with the json in file: mocks/mock-bip-claims-api/src/main/resources/mock-special-issues.json*/
```

File: svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiProps.java [hunk 5]
```diff
@@ -49,6 +51,11 @@ public String getClaimRequestUrl(String path) {
     }
     return baseUrl + path;
   }
+
+  public String getAvailabilityUrl() {
```
Comment (dfitchett) [inline svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiProps.java:55]: Probably dont need this here if you remove the `ServiceBipApi.isApiFunctioning()` method.

File: svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiService.java [hunk 6]
```diff
@@ -224,8 +231,8 @@ private <T extends BipPayloadResponse> T makeRequest(
    * @return true if the API responds with OK status and response.
    */
   public boolean isApiFunctioning() {
```
Comment (dfitchett) [inline svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiService.java:241]: I'd say this method is unnecessary and can be removed. It's only used in test.

File: svc-bip-api/src/test/java/gov/va/vro/bip/service/BipApiServiceTest.java
Comment (dfitchett) [inline svc-bip-api/src/test/java/gov/va/vro/bip/service/BipApiServiceTest.java:1]: If these don't need to be package private, this can be reverted.

**Review by dfitchett — COMMENTED** [review]
You should remove commented code / comments!

File: q [hunk 7]
```diff
@@ -0,0 +1,7 @@
+
+# Nix
+# if [ -e '/nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh' ]; then
+#   . '/nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh'
+# fi
+# End Nix
+
```
Comment (lisac) [inline q]: is this file needed for this PR? i don't see it referenced. could we add a note/comment on how to use it?

File: svc-bip-api/src/integrationTest/java/gov/va/vro/bip/SpecialIssueTypesTest.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  //   bipApiProps.setClaimBaseUrl(BipApiServiceTest.CLAIM_URL);
+  //   bipApiProps.setClaimIssuer(BipApiServiceTest.CLAIM_ISSUER);
+  //   bipApiProps.setClaimSecret(BipApiServiceTest.CLAIM_SECRET);
+  //   bipApiProps.setApplicationId(BipApiServiceTest.APP_ID);
+  //   bipApiProps.setStationId(BipApiServiceTest.STATION_ID);
+  //   bipApiProps.setClaimClientId(BipApiServiceTest.CLAIM_USERID);
+
+  //   service = new SpecialIssueStateService(restTemplate, bipApiProps, mapper,
+  // metricLoggerService);
+  // }
```
Comment (lisac) [inline svc-bip-api/src/integrationTest/java/gov/va/vro/bip/SpecialIssueTypesTest.java]: just to double-check: do we want to keep this `setUp()` commented out (versus deleting)?

File: svc-bip-api/src/main/java/gov/va/vro/bip/config/BipApiConfig.java [hunk 9]
```diff
@@ -55,6 +55,10 @@ public class BipApiConfig {
   public BipApiProps getBipApiProps() {
     return new BipApiProps();
   }
+  // @Bean
+  // public BipApiProps getBipSpecialInterestService() {
+  //   return new SpecialIssueStateService();
+  // }
```
Comment (lisac) [inline svc-bip-api/src/main/java/gov/va/vro/bip/config/BipApiConfig.java]: potential to delete this commented-out code?

File: svc-bip-api/src/test/java/gov/va/vro/bip/service/BipApiServiceTest.java [hunk 10]
```diff
@@ -62,33 +62,30 @@
...
+  static final String CANCEL_CLAIM = "/claims/%s/cancel";
+  static final String CONTENTION = "/claims/%s/contentions";
+  static final String TEMP_STATION_OF_JURISDICTION = "/claims/%s/temporary_station_of_jurisdiction";
+  static final String HTTPS = "https://";
+  static final String CLAIM_URL = "claims.bip.va.gov";
+  static final String CLAIM_SECRET = "thesecretissecurenowthatitislongenough2184vnrwma";
+  static final String CLAIM_USERID = "userid";
+  static final String CLAIM_ISSUER = "issuer";
+  static final String STATION_ID = "280";
+  static final String APP_ID = "bip";
```
Comment (lisac) [inline svc-bip-api/src/test/java/gov/va/vro/bip/service/BipApiServiceTest.java]: question: is this removal of the `private` access modifier linked to a particular need? or is it a matter of general code quality?

**Review by lisac — APPROVED** [review]

File: mocks/mock-bip-claims-api/src/main/java/gov/va/vro/mockbipclaims/model/bip/response/SpecialIssueTypesResponse.java (hunk 1 above)
Comment (nelsestu) [reply mocks/mock-bip-claims-api/src/main/java/gov/va/vro/mockbipclaims/model/bip/response/SpecialIssueTypesResponse.java:10]: @dfitchett Question about the removing commented code comment.  Removing this specific comment introduces a new compiler warning since all public members require a javadoc comment.  I am interested in your thoughts about that case.  I would personally prefer to keep the comment and reduce our compiler warnings.  In generally though, i agree that useless comments are better to remove.

File: svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/GetSpecialIssueTypesResponse.java (hunk 2 above)
Comment (nelsestu) [reply svc-bip-api/src/main/java/gov/va/vro/bip/model/contentions/GetSpecialIssueTypesResponse.java]: @brostk, No, no convention on single line ifs.  I am surprised that my IDE didn't fix that, but I am even more surprised that the spotless java style check allowed it. i'll go ahead with the fix so that we don't end up with more of these single lines.

File: svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiService.java (hunk 6 above)
Comment (nelsestu) [reply svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiService.java:241]: Removal of isApiFunctioning would require a lot more significant changes, and since i didn't originate that function and there are unit tests, I'm going to leave it alone.  The one thing I did do was modified its implementation, so I'll go ahead and modify it back to what it was.

File: svc-bip-api/src/integrationTest/java/gov/va/vro/bip/SpecialIssueTypesTest.java (hunk 8 above)
Comment (nelsestu) [reply svc-bip-api/src/integrationTest/java/gov/va/vro/bip/SpecialIssueTypesTest.java]: no we definitely prefer to delete it :)

File: q (hunk 7 above)
Comment (nelsestu) [reply q]: oh i see, "q" is the Amazon Codewhisperer agent, so yes, we'll still remove it

File: q
Comment (nelsestu) [reply q:1]: that is a mystery, but has something to do with amazon's CodeWhisperer agent.  naturally it has been removed.

File: svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiService.java (hunk 6 above)
Comment (nelsestu) [reply svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiService.java:241]: Actually, I am recalling now that this function isApiFunctioning was added to allow for quick test cycles of checking the mutual TLS authentication with the BIP api. While it is "only" used in unit tests, we were ultimately able to get this test running on a windows based citrix environment.  It still doesn't seem feasible to get more than this one test running in a citrix environment, so that makes the isApiFunctioning unit tests still a valuable troubleshooting prospect.

File: svc-bip-api/src/test/java/gov/va/vro/bip/service/BipApiServiceTest.java (hunk 10 above)
Comment (nelsestu) [reply svc-bip-api/src/test/java/gov/va/vro/bip/service/BipApiServiceTest.java]: the private access modifier was removed originally so that these variables could be used in multiple test fixtures. Ultimately the additional fixture was removed when I dropped the special issue code lookup functionality, so now I've reintroduced the private modifier.

File: svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiProps.java (hunk 5 above)
Comment (nelsestu) [reply svc-bip-api/src/main/java/gov/va/vro/bip/service/BipApiProps.java:55]: I've decided to keep this function, because it helps clarify the use case this URL has. First, I would argue that isApiFunctioning is a valuable troubleshooting element.  For more details about why please see: https://github.com/department-of-veterans-affairs/abd-vro/pull/3280#discussion_r1716059336.  If we accept that isApiFunctioning has value, the original implementation was using the special_issue_types endpoint to validate the request details to the BIP API were functioning properly.  By splitting out a function called getAvailabilityUrl and returning special_issue_types url in that function, it helps formalize the use of that endpoint for this purpose. it is just more explicit to keep the function, and there is no downsides that i can see.

File: svc-bip-api/src/main/java/gov/va/vro/bip/config/BipApiConfig.java (hunk 9 above)
Comment (nelsestu) [reply svc-bip-api/src/main/java/gov/va/vro/bip/config/BipApiConfig.java]: removed

**Review by dfitchett — APPROVED** [review]

**Review by nelsestu — COMMENTED** [review]
Thanks @dfitchett , I wasn't understanding how the unrelated integration test would be failing, but removing the extra RestTemplate mock that was being created certainly makes sense in hindsight.
